docs/references/http-gateway-protocol-spec.md

```diff
@@ -0,0 +1,471 @@
+- This is unexpected for developers and will lead to vulnerabilities in dApps sooner or later.
+- The effectiveness of security headers (such as Content Security Policy) is diminished as they can be omitted or modified by malicious nodes.
+
+[Response Verification version 2](#response-verification) overcomes these issues.
+
+The steps for response verification are as follows:
+
+- See the [response verification outline](#response-verification-outline) for the full subprotocol description.
+- The path `["http_assets", <url>]` exists in the `tree` and is a leaf with a value, where `<url>` is the utf8-encoded URL from the `HttpRequest`.
+- Otherwise, the path `["http_assets", "/index.html"]` must exist in the `tree` and be a leaf.
+- That leaf must contain the SHA-256 hash of the decoded body.
+  - If the `streaming_strategy` field of the `HttpResponse` is set, all chunks are streamed and concatenated according to [response body streaming](#response-body-streaming) before decoding.
+  - The body is decoded according to the `Content-Encoding` header if present.
+  - Supported values for the `Content-Encoding` header include `gzip` and `deflate`.
+
+## Canister HTTP Interface
+
+The full [Candid](https://github.com/dfinity/candid/blob/master/spec/Candid.md) interface that a canister is expected to implement is as follows:
+
+```
+type HeaderField = record { text; text; };
+
+type HttpRequest = record {
+    method: text;
+    url: text;
+    headers: vec HeaderField;
+    body: blob;
+    certificate_version: opt nat;
+};
+
+type HttpResponse = record {
+    status_code: nat16;
+    headers: vec HeaderField;
+    body: blob;
+    upgrade : opt bool;
+    streaming_strategy: opt StreamingStrategy;
+};
+
+// Each canister that uses the streaming feature gets to choose their concrete
+// type; the HTTP Gateway will treat it as an opaque value that is only fed to
+// the callback method
+
+type StreamingToken = /* application-specific type */
+
+type StreamingCallbackHttpResponse = record {
+    body: blob;
+    token: opt StreamingToken;
+};
+
+type StreamingStrategy = variant {
+    Callback: record {
+        callback: func (StreamingToken) -> (opt StreamingCallbackHttpResponse) query;
+        token: StreamingToken;
+    };
+};
+
+service : {
+    http_request: (request: HttpRequest) -> (HttpResponse) query;
+    http_request_update: (request: HttpRequest) -> (HttpResponse);
+}
+```
+
+Not all of this interface is required. The following sections detail what can be optionally omitted depending on the requirements of the canister in question.
+
+### Response Verification Interface
+
+The `certificate_version` field of the `HttpRequest` interface is optional depending on the version of [response verification](#response-verification) that the canister is implementing. It is omitted in older canisters that do not implement response verification version 2 or later.
```

Comment:
I don't understand this paragraph. Doesn't the gateway set the `certificate_version` field in the request? The older canisters will just not know what to do with it.

Reply:
Nevermind. I got it. The canister needs to implement the interface. Sorry!